### PR TITLE
[DPE-6043] - Add wrappers and rename daemon

### PR DESCRIPTION
**Pull request by skourta**
This pull request includes several changes to improve the configuration and security of the `etcd` snap package. The most important changes involve updating configuration paths, adding scripts for privilege dropping, and modifying the `snapcraft.yaml` file to enhance security and functionality.

Configuration updates:

* [`config/etcd.conf.yml`](diffhunk://#diff-11b9dce02506e4a04e90ec8fd6ae5a1303c63b6d2f598c27a38c614233945022L7-R7): Updated the `data-dir` path to use the snap's common directory.

Security enhancements:

* [`snap/hooks/install`](diffhunk://#diff-e58b9a041906b37e03356984486699655b2d559fa8dc4d95b340a3b5be7a9a42R6-R18): Added commands to set up permissions and ownership for the snap's data directory.
* [`snap/local/drop_priv.sh`](diffhunk://#diff-73c43d081f46ddf34bf5af53db6b4768c0218d4fad21def464618f5b757a7821R1-R17): Added a new script to run applications with restricted privileges using `setpriv`.
* [`snap/local/start-etcd.sh`](diffhunk://#diff-6d0a469ae28efb2b2496d4c03e9e615ceb759f1895d764e17c0aced349ddfcb2R1-R4): Added a new script to run `etcd` as a non-sudo user for security reasons.

Functionality improvements:

* [`snap/snapcraft.yaml`](diffhunk://#diff-56759910381a014fecfd7556dd72ddd68c747d922a5b7df2044b9ce7c552f5f5R12-R42): Updated the `apps` section to use the new `start-etcd.sh` script for the `etcd` command, added `system-usernames`, and modified the `etcdctl` and `etcdutl` commands to use the `drop_priv.sh` script. [[1]](diffhunk://#diff-56759910381a014fecfd7556dd72ddd68c747d922a5b7df2044b9ce7c552f5f5R12-R42) [[2]](diffhunk://#diff-56759910381a014fecfd7556dd72ddd68c747d922a5b7df2044b9ce7c552f5f5R55-R57)

### CODE REVIEW

File: snap/hooks/install [hunk 1]
```diff
@@ -2,4 +2,17 @@
 
 # This hook is called when the snap is installed
 # copy the configuration file to the snap's data directory
-cp $SNAP/config/etcd.conf.yml $SNAP_DATA/etcd.conf.yml
\ No newline at end of file
+cp $SNAP/config/etcd.conf.yml $SNAP_DATA/etcd.conf.yml
```
Comment (Mehdi-Bendriss) [inline snap/hooks/install]:
```suggestion
cp "${SNAP}"/config/etcd.conf.yml "${SNAP_DATA}"/etcd.conf.yml
```

File: snap/hooks/install [hunk 2]
```diff
@@ -2,4 +2,17 @@
...
+chown -R 584788:584788 "${SNAP_DATA}"/*
+
+# create the data directory
+export DATA="${SNAP_COMMON}"/var/lib/etcd
+mkdir -p "${DATA}"
+
+# If we just created the directory, we set up permissions
+if [ stat -c '%u' "$DATA" == 0 ]; then
```
Comment (Mehdi-Bendriss) [inline snap/hooks/install]:
```suggestion
if [ stat -c '%u' "${DATA}" == 0 ]; then
```

File: snap/hooks/install [hunk 3]
```diff
@@ -2,4 +2,17 @@
 
 # This hook is called when the snap is installed
 # copy the configuration file to the snap's data directory
-cp $SNAP/config/etcd.conf.yml $SNAP_DATA/etcd.conf.yml
\ No newline at end of file
+cp $SNAP/config/etcd.conf.yml $SNAP_DATA/etcd.conf.yml
+chown -R 584788:584788 "${SNAP_DATA}"/*
+
+# create the data directory
+export DATA="${SNAP_COMMON}"/var/lib/etcd
```
Comment (Mehdi-Bendriss) [inline snap/hooks/install]: I think there is no need to `export` here
```suggestion
DATA="${SNAP_COMMON}"/var/lib/etcd
```

File: snap/hooks/install [hunk 4]
```diff
@@ -2,4 +2,17 @@
...
+mkdir -p "${DATA}"
+
+# If we just created the directory, we set up permissions
+if [ stat -c '%u' "$DATA" == 0 ]; then
+    chmod -R 770 "${SNAP_COMMON}"
+    chmod 750 "${DATA}"
+fi
+
+chown -R 584788:584788 "${SNAP_COMMON}"/*
```
Comment (Mehdi-Bendriss) [inline snap/hooks/install]: nit: please keep an empty line at the end of source files 
```suggestion
chown -R 584788:584788 "${SNAP_COMMON}"/*

```

File: snap/local/drop_priv.sh [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+        --clear-groups \
+        --reuid snap_daemon \
+        --regid snap_daemon \
+        -- \
+        "${SNAP}/bin/$@"
+else
+    exec "${SNAP}/bin/$@"
+fi
+
+popd > /dev/null
```
Comment (Mehdi-Bendriss) [inline snap/local/drop_priv.sh]: nit:
```suggestion
popd > /dev/null

```

File: snap/local/drop_priv.sh [hunk 6]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
Comment (Mehdi-Bendriss) [inline snap/local/drop_priv.sh]: maybe better:
```suggestion
#!/usr/bin/env bash
```

File: snap/local/drop_priv.sh [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+# Wrapper script for charmed mongodb applications to be run with restricted privileges
```
Comment (Mehdi-Bendriss) [inline snap/local/drop_priv.sh]:
```suggestion
# Wrapper script for charmed ETCD to be run with restricted privileges
```

File: snap/local/start-etcd.sh [hunk 8]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
```
Comment (Mehdi-Bendriss) [inline snap/local/start-etcd.sh]:
```suggestion
#!/usr/bin/env bash
```

File: snap/local/start-etcd.sh [hunk 9]
```diff
@@ -0,0 +1,4 @@
...
+
+# For security measures, daemons should not be run as sudo. Execute mongod as the non-sudo user: snap-daemon.
+exec $SNAP/usr/bin/setpriv --clear-groups --reuid snap_daemon --regid snap_daemon -- $SNAP/bin/etcd
```
Comment (Mehdi-Bendriss) [inline snap/local/start-etcd.sh]:
```suggestion
exec "${SNAP}"/usr/bin/setpriv --clear-groups --reuid snap_daemon --regid snap_daemon -- "${SNAP}"/bin/etcd

```

**Review by Mehdi-Bendriss — COMMENTED** [review]
Thanks Smail, looks good! I have a few minor comments

File: snap/local/start-etcd.sh [hunk 10]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+
+# For security measures, daemons should not be run as sudo. Execute mongod as the non-sudo user: snap-daemon.
```
Comment (Mehdi-Bendriss) [inline snap/local/start-etcd.sh]:
```suggestion
# For security measures, daemons should not be run as sudo. Execute ETCD as the non-sudo user: snap-daemon.
```

**Review by Mehdi-Bendriss — APPROVED** [review]
Great work!

**Review by reneradoi — APPROVED** [review]
No further comments from my side. Thank you @skourta !

**Pull request merged**